src/Spago/Command/Build.purs

```diff
@@ -140,56 +96,188 @@ run opts = do
+      logDebug $ "Getting purs graph using globs for package: " <> PackageName.print selected.package.name
+      maybeGraph <- Graph.runGraph globs opts.pursArgs
+      case maybeGraph of
+        Nothing ->
+          logWarn $
+            "Due to JSON decoding failure on 'purs graph' output, \
+            \packages that define modules with the same name might not be detected."
+        Just graph -> do
+          let
+            toModulesDefinedByThisPackage :: Purs.ModuleGraphNode -> Maybe (Array (Tuple PackageName FilePath))
+            toModulesDefinedByThisPackage v =
+              [ Tuple selected.package.name v.path ] <$ (String.stripPrefix (String.Pattern selected.path) v.path)
+            modulesDefinedByThisPackage = Map.mapMaybe toModulesDefinedByThisPackage $ unwrap graph
+
+          liftEffect $ Ref.modify_ (Map.unionWith (<>) modulesDefinedByThisPackage) duplicateModuleRef
+          when workspace.buildOptions.pedanticPackages do
+            logInfo $ "Looking for unused and undeclared transitive dependencies..."
+            env <- ask
+            errors <- Graph.toImportErrors selected <$> runSpago (Record.union { graph, selected } env) Graph.checkImports
+            unless (Array.null errors) do
+              die' errors
+
+    else do
+      when workspace.buildOptions.pedanticPackages do
+        logInfo $ "Looking for unused and undeclared transitive dependencies..."
+        maybeGraph <- Graph.runGraph globs opts.pursArgs
+        for_ maybeGraph \graph -> do
+          env <- ask
+          errors <- Graph.toImportErrors selected <$> runSpago (Record.union { graph, selected } env) Graph.checkImports
+          unless (Array.null errors) do
+            die' errors
+
+  when buildingMultiplePackages do
+    moduleMap <- liftEffect $ Ref.read duplicateModuleRef
+    let duplicateModules = Map.filter (\packages -> Array.length packages > 1) moduleMap
+    unless (Map.isEmpty duplicateModules) do
+      die $
+        ( Dodo.lines
+            [ Dodo.text $ "Detected " <> show (Map.size duplicateModules) <> " modules with the same module name across 2 or more packages defined in this workspace."
+            , duplicateModules
+                # Map.toUnfoldable
+                # Array.mapWithIndex
+                    ( \idx (Tuple m ps) ->
+                        Dodo.lines
+                          [ Dodo.text $ show (idx + 1) <> ") Module \"" <> m <> "\" was defined in the following packages:"
+                          , Dodo.indent
+                              $ Dodo.lines
+                              $ map
+                                  ( \(Tuple pkg moduleFilePath) ->
+                                      Dodo.text $ "- " <> PackageName.print pkg <> "   at path: " <> moduleFilePath
+                                  )
+                              $ Array.sort ps
+                          ]
+                    )
+                # Dodo.lines
+            ] :: Docc
+        )
 
```

Comment:
Let's extract this formatting stuff to a separate function

Reply:
Done.